### PR TITLE
MacOS: Fix memory leak duing initialization

### PR DESCRIPTION
**Pull request by nl78**
Low priority :-)

### CODE REVIEW

File: src/grisbi_osx.m [hunk 1]
```diff
@@ -58,19 +58,28 @@
...
 }
 
-static char const *get_program_dir()
+static gchar *get_program_dir()
 {
-    return g_path_get_dirname(get_program_name());
+    char *program_name = get_program_name();
+    gchar *program_dir = g_path_get_dirname(program_name);
+    free(program_name);
+    return program_name;
```
Comment (LudovicRousseau) [inline src/grisbi_osx.m:66]: I am not sure that returning `program_name` that was `free()` just above is a good idea.
Maybe you mean `return program_dir;`?

**Review by LudovicRousseau — CHANGES_REQUESTED** [review]

File: src/grisbi_osx.m (hunk 1 above)
Comment (nl78) [reply src/grisbi_osx.m]: You're right! :)
FIXED in last update

File: src/grisbi_osx.m [hunk 2]
```diff
@@ -58,19 +58,28 @@
...
     if (g_str_has_suffix(program_dir, "Contents/MacOS")) {
-        prefix = g_build_filename(prefix, "Resources", NULL);
+        get_bundle_prefix = g_build_filename(prefix, "Resources", NULL);
+    } else {
+        get_bundle_prefix = g_strdup(prefix);
     }
 
+    g_free(program_dir);
+    g_free(prefix);
+
```
Comment (LudovicRousseau) [inline src/grisbi_osx.m:82]: Do not `g_free(prefix)` just before you return it.

File: src/grisbi_osx.m [hunk 3]
```diff
@@ -108,36 +117,68 @@ static void my_setenv(const gchar *key, const gchar *value) {
...
+    g_free(support_share_dir);
+
     my_setenv("XDG_DATA_DIRS",   bundle_resources_share_dir);
-    my_setenv("XDG_CONFIG_HOME", g_build_filename(app_support_dir, "config", NULL));
-    my_setenv("XDG_CONFIG_DIRS", g_build_filename(bundle_resources_etc_dir, "xdg", NULL));
-    my_setenv("XDG_CACHE_HOME",  g_build_filename(app_support_dir, "cache", NULL));
+
+    gchar *support_config_dir = g_build_filename(app_support_dir, "config", NULL);
+    my_setenv("XDG_CONFIG_HOME", support_config_dir);
+    g_free(support_config_dir);
```
Comment (LudovicRousseau) [inline src/grisbi_osx.m:128]: I suggest to create a macro that would do something like:
`SETENV("XDG_CONFIG_HOME", g_build_filename(app_support_dir, "config", NULL));`

With:
`#define SETENV(variable, value) do {gchar *tmp = value; my_setenv(variable, tmp); g_free(tmp);} while(0)`
The idea is to remove code duplication and simplify code reading/understanding.

The code of `my_setenv()` could also be moved inside the macro.

**Review by LudovicRousseau — CHANGES_REQUESTED** [review]

File: src/grisbi_osx.m (hunk 3 above)
Comment (nl78) [reply src/grisbi_osx.m:128]: nope, some values may be commo for few variables.

File: src/grisbi_osx.m (hunk 2 above)
Comment (nl78) [reply src/grisbi_osx.m:82]: I means return bundle_prefix.... FIXED

File: src/grisbi_osx.m [hunk 4]
```diff
@@ -108,36 +117,68 @@ static void my_setenv(const gchar *key, const gchar *value) {
...
     // DYLD_LIBRARY_PATH
     // This is required to make Python GTK bindings work as they use dlopen()
     // to load libraries.
-    my_setenv("DYLD_LIBRARY_PATH", g_build_path(":", bundle_resources_lib_dir, NULL));
+    gchar *dyld_path_dir = g_build_path(":", bundle_resources_lib_dir, NULL);
+    my_setenv("DYLD_LIBRARY_PATH", dyld_path_dir);
+    g_free(dyld_path_dir);
+
+    g_free(app_support_dir);
+    g_free(bundle_resources_dir);
```
Comment (LudovicRousseau) [inline src/grisbi_osx.m:177]: Do not `g_free(bundle_resources_dir);` if you return `bundle_resources_dir`

**Review by LudovicRousseau — CHANGES_REQUESTED** [review]

File: src/grisbi_osx.m [hunk 5]
```diff
@@ -58,20 +58,29 @@
     return program_name;
 }
 
-static char const *get_program_dir()
+static gchar *get_program_dir()
```
Comment (LudovicRousseau) [inline src/grisbi_osx.m:61]: I like to have `get_program_dir(void)` when the function takes no parameter.

Comment (nl78) [reply src/grisbi_osx.m:61]: désolé c'est poussif, j'aurai du relire mon code :)

**Comment by LudovicRousseau** [comment]
Merged

**Comment by nl78** [comment]
C'est tout bon!

**Pull request closed without merging**